### PR TITLE
Publish OCI image and release Kubernetes resources

### PR DESCRIPTION
**Pull request by baijum**
Signed-off-by: Baiju Muthukadan <baiju.m.mail@gmail.com>

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -76,3 +76,9 @@ deploy: manifests ## Deploy controller to the K8s cluster specified in ~/.kube/c
 
 undeploy: ## Undeploy controller from the K8s cluster specified in ~/.kube/config.
 	$(KUSTOMIZE) build config/default | $(KO) delete -f -
+
+##@ Publish
+
+publish:
+	echo "${AUTH_TOKEN}" | $(KO) login ghcr.io --username ${REPOSITORY_OWNER} --password-stdin
```
Comment (scothis) [inline Makefile:83]: Auth to a registry should be part of the workflow, not the makefile

```suggestion
```

File: Makefile [hunk 2]
```diff
@@ -76,3 +76,9 @@ deploy: manifests ## Deploy controller to the K8s cluster specified in ~/.kube/c
...
+	$(KO) publish -B --tags ${RELEASE_TAG} ./
```
Comment (scothis) [inline Makefile:84]: `${RELEASE_TAG}` should have a default specified by the Makefile.

A common pattern is to define the variable with a default. Something like:

```
VERSION ?= "dev"
```

File: Makefile [hunk 3]
```diff
@@ -76,3 +76,9 @@ deploy: manifests ## Deploy controller to the K8s cluster specified in ~/.kube/c
 
 undeploy: ## Undeploy controller from the K8s cluster specified in ~/.kube/config.
 	$(KUSTOMIZE) build config/default | $(KO) delete -f -
+
+##@ Publish
+
+publish:
```
Comment (scothis) [inline Makefile:82]: add a comment for the target, so that it shows up in `make help`

**Review by scothis — COMMENTED** [review]
The publish make target is only publishing an image, which seems strange. I'd expect the output to be a yaml file that references an image.

File: Makefile (hunk 1 above)
Comment (baijum) [reply Makefile]: Done. 240399d51dca528f3f5341dc5e629473ed34415d

File: Makefile (hunk 2 above)
Comment (baijum) [reply Makefile]: Done. 240399d51dca528f3f5341dc5e629473ed34415d

File: Makefile (hunk 3 above)
Comment (baijum) [reply Makefile]: Done. 240399d51dca528f3f5341dc5e629473ed34415d

File: config/manager/manager.yaml [hunk 4]
```diff
@@ -27,7 +27,7 @@ spec:
       containers:
       - args:
         - --leader-elect
-        image: ko://github.com/k8s-service-bindings/implementation
+        image: ko://github.com/k8s-service-bindings/service-binding-controller
```
Comment (baijum) [inline config/manager/manager.yaml:30]: This is an unrelated change. Let me know if I should create a separate PR.

File: hack/ko/go.mod [hunk 5]
```diff
@@ -1,4 +1,4 @@
-module github.com/k8s-service-bindings/implementation/hack/ko
+module github.com/k8s-service-bindings/service-binding-controller/hack/ko
```
Comment (baijum) [inline hack/ko/go.mod:1]: This is an unrelated change. Let me know if I should create a separate PR.

File: Makefile [hunk 6]
```diff
@@ -76,3 +77,8 @@ deploy: manifests ## Deploy controller to the K8s cluster specified in ~/.kube/c
 
 undeploy: ## Undeploy controller from the K8s cluster specified in ~/.kube/config.
 	$(KUSTOMIZE) build config/default | $(KO) delete -f -
+
+##@ Publish
+
+publish: ## Publish OCI image and prepare Kubernetes resources for release
```
Comment (baijum) [inline Makefile:83]: Not sure if the publish target has much use outside the CI. Let me know if we should move this command to the CI configuration.

Comment (scothis) [reply Makefile:83]: I had the same thought. The difference between publish and deploy is that publish doesn't apply the config to the cluster.

File: .github/workflows/ci.yaml [hunk 7]
```diff
@@ -22,3 +22,21 @@ jobs:
...
+      env:
+        KO_DOCKER_REPO: ghcr.io/${{ github.repository_owner }}
+        AUTH_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+        REPOSITORY_OWNER: ${{ github.repository_owner }}
+        VERSION: ${{ steps.version.outputs.tag }}
+      run: |
+        echo "${AUTH_TOKEN}" | go run -modfile hack/ko/go.mod github.com/google/ko login ghcr.io --username ${REPOSITORY_OWNER} --password-stdin
+        make publish
+    - name: Update the release with Kubernetes resources
+      uses: softprops/action-gh-release@v0.1.12
```
Comment (scothis) [inline .github/workflows/ci.yaml:39]: Let's use the official actions to create a release and attach a file to the release.

File: .github/workflows/ci.yaml [hunk 8]
```diff
@@ -22,3 +22,21 @@ jobs:
...
+      run: echo ::set-output name=tag::${GITHUB_REF/refs\/tags\//}
+    - name: Publish
+      if: startsWith(github.ref, 'refs/tags/')
+      env:
+        KO_DOCKER_REPO: ghcr.io/${{ github.repository_owner }}
+        AUTH_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+        REPOSITORY_OWNER: ${{ github.repository_owner }}
+        VERSION: ${{ steps.version.outputs.tag }}
+      run: |
+        echo "${AUTH_TOKEN}" | go run -modfile hack/ko/go.mod github.com/google/ko login ghcr.io --username ${REPOSITORY_OWNER} --password-stdin
```
Comment (scothis) [inline .github/workflows/ci.yaml:36]:
```suggestion
        echo "${AUTH_TOKEN}" | docker login ghcr.io --username ${REPOSITORY_OWNER} --password-stdin
```

File: .github/workflows/ci.yaml (hunk 7 above)
Comment (baijum) [reply .github/workflows/ci.yaml:39]: This official once looks archived and recommend this one: https://github.com/actions/upload-release-asset

File: config/manager/manager.yaml (hunk 4 above)
Comment (baijum) [reply config/manager/manager.yaml:30]: PR: https://github.com/k8s-service-bindings/service-binding-controller/pull/26

File: hack/ko/go.mod (hunk 5 above)
Comment (baijum) [reply hack/ko/go.mod:1]: PR: https://github.com/k8s-service-bindings/service-binding-controller/pull/26

File: config/manager/manager.yaml (hunk 4 above)
Comment (scothis) [reply config/manager/manager.yaml:30]: By 👍 I meant that the unrelated change was ok, but I merged #26 since it was ready to go :)

File: .github/workflows/ci.yaml (hunk 7 above)
Comment (scothis) [reply .github/workflows/ci.yaml:39]: today I learned...

File: .github/workflows/ci.yaml [hunk 9]
```diff
@@ -22,3 +22,21 @@ jobs:
...
+    - name: Publish
+      if: startsWith(github.ref, 'refs/tags/')
+      env:
+        KO_DOCKER_REPO: ghcr.io/${{ github.repository_owner }}
+        AUTH_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+        REPOSITORY_OWNER: ${{ github.repository_owner }}
+        VERSION: ${{ steps.version.outputs.tag }}
+      run: |
+        echo "${AUTH_TOKEN}" | docker login ghcr.io --username ${REPOSITORY_OWNER} --password-stdin
+        go run -modfile hack/kustomize/go.mod sigs.k8s.io/kustomize/kustomize/v3 build config/default | go run -modfile hack/ko/go.mod github.com/google/ko resolve -B --tags ${VERSION} -f - > release.yaml
```
Comment (scothis) [inline .github/workflows/ci.yaml:37]: attempting to improve readability, your call to take or leave it

```suggestion
      env:
        GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
        KO: go run -modfile hack/ko/go.mod github.com/google/ko
        KO_DOCKER_REPO: ghcr.io/${{ github.repository }}
        KUSTOMIZE: go run -modfile hack/kustomize/go.mod sigs.k8s.io/kustomize/kustomize/v3
        VERSION: ${{ steps.version.outputs.tag }}
      run: |
        echo "${GITHUB_TOKEN}" | docker login ghcr.io --username ${{ github.repository_owner }} --password-stdin
        ${KUSTOMIZE} build config/default | ${KO} resolve --tags "${VERSION}" -f - > release.yaml
```

**Comment by baijum** [comment]
I tested this in another sandbox org. This is how it looks like: https://github.com/sbc-sandbox/service-binding-controller/pkgs/container/service-binding-controller

Release: https://github.com/sbc-sandbox/service-binding-controller/releases/tag/v0.1.4 (I changed the draft after release)

I think we can merge this (need to squash the commits).

File: .github/workflows/ci.yaml [hunk 10]
```diff
@@ -40,5 +40,6 @@ jobs:
       uses: softprops/action-gh-release@v0.1.12
       if: startsWith(github.ref, 'refs/tags/')
       with:
+        tag_name: ${{ steps.version.outputs.tag }}
```
Comment (scothis) [inline .github/workflows/ci.yaml:43]: isn't that the default?

Comment (baijum) [reply .github/workflows/ci.yaml:43]: Yes, I will remove this field.
